#include "common_dbc.h"

namespace {

const Signal sigs_128[] = {
    {
      .name = "WHEELSPEED_FR",
      .b1 = 0,
      .b2 = 16,
      .bo = 48,
      .is_signed = false,
      .factor = 0.034,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "WHEELSPEED_FL",
      .b1 = 16,
      .b2 = 16,
      .bo = 32,
      .is_signed = false,
      .factor = 0.034,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "WHEELSPEED_RR",
      .b1 = 32,
      .b2 = 16,
      .bo = 16,
      .is_signed = false,
      .factor = 0.034,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "WHEELSPEED_RL",
      .b1 = 48,
      .b2 = 16,
      .bo = 0,
      .is_signed = false,
      .factor = 0.034,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_144[] = {
    {
      .name = "BRAKEPEDAL",
      .b1 = 2,
      .b2 = 1,
      .bo = 61,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "ESP_STATUS",
      .b1 = 6,
      .b2 = 1,
      .bo = 57,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "VEHICLESPEED",
      .b1 = 42,
      .b2 = 6,
      .bo = 16,
      .is_signed = true,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_194[] = {
    {
      .name = "SMART_STEER_ANGLE_VALUE",
      .b1 = 0,
      .b2 = 15,
      .bo = 49,
      .is_signed = false,
      .factor = 0.04375,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SMART_STEER_ANGLE_SIGN",
      .b1 = 15,
      .b2 = 1,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SMART_STEER_RATE_VALUE",
      .b1 = 16,
      .b2 = 15,
      .bo = 33,
      .is_signed = false,
      .factor = 0.04375,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SMART_STEER_RATE_SIGN",
      .b1 = 31,
      .b2 = 1,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_400[] = {
    {
      .name = "GEAR",
      .b1 = 2,
      .b2 = 5,
      .bo = 57,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "RPM",
      .b1 = 16,
      .b2 = 7,
      .bo = 41,
      .is_signed = false,
      .factor = 50,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_544[] = {
    {
      .name = "RIGHT_DOOR",
      .b1 = 0,
      .b2 = 1,
      .bo = 63,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "RIGHT_SIGNAL",
      .b1 = 4,
      .b2 = 1,
      .bo = 59,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "LEFT_SIGNAL",
      .b1 = 5,
      .b2 = 1,
      .bo = 58,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SIDELIGHTS",
      .b1 = 6,
      .b2 = 1,
      .bo = 57,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "LEFT_DOOR",
      .b1 = 7,
      .b2 = 1,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_784[] = {
    {
      .name = "THROTTLE_POSITION",
      .b1 = 32,
      .b2 = 16,
      .bo = 16,
      .is_signed = false,
      .factor = 0.00156143,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};

const Msg msgs[] = {
  {
    .name = "SMARTROADSTERWHEELSPEEDS",
    .address = 0x80,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_128),
    .sigs = sigs_128,
  },
  {
    .name = "ABS",
    .address = 0x90,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_144),
    .sigs = sigs_144,
  },
  {
    .name = "SMARTSTEERINGANGLESENSOR",
    .address = 0xC2,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_194),
    .sigs = sigs_194,
  },
  {
    .name = "GEAR_PACKET",
    .address = 0x190,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_400),
    .sigs = sigs_400,
  },
  {
    .name = "BODYCONTROL",
    .address = 0x220,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_544),
    .sigs = sigs_544,
  },
  {
    .name = "SMARTTHROTTLEPEDAL",
    .address = 0x310,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_784),
    .sigs = sigs_784,
  },
};

const Val vals[] = {
    {
      .name = "BRAKEPEDAL",
      .address = 0x90,
      .def_val = "0 NOT_PRESSED 1 PRESSED",
      .sigs = sigs_144,
    },
    {
      .name = "ESP_STATUS",
      .address = 0x90,
      .def_val = "0 ON 1 OFF",
      .sigs = sigs_144,
    },
    {
      .name = "GEAR",
      .address = 0x190,
      .def_val = "0 P 2 N 1 R 19 D 3 S",
      .sigs = sigs_400,
    },
    {
      .name = "LEFT_DOOR",
      .address = 0x220,
      .def_val = "0 OPEN 1 CLOSED",
      .sigs = sigs_544,
    },
    {
      .name = "LEFT_SIGNAL",
      .address = 0x220,
      .def_val = "0 OFF 1 ON",
      .sigs = sigs_544,
    },
    {
      .name = "RIGHT_DOOR",
      .address = 0x220,
      .def_val = "0 OPEN 1 CLOSED",
      .sigs = sigs_544,
    },
    {
      .name = "RIGHT_SIGNAL",
      .address = 0x220,
      .def_val = "0 OFF 1 ON",
      .sigs = sigs_544,
    },
    {
      .name = "SIDELIGHTS",
      .address = 0x220,
      .def_val = "0 OFF 1 ON",
      .sigs = sigs_544,
    },
};

}

const DBC ocelot_smart_roadster_pt = {
  .name = "ocelot_smart_roadster_pt",
  .num_msgs = ARRAYSIZE(msgs),
  .msgs = msgs,
  .vals = vals,
  .num_vals = ARRAYSIZE(vals),
};

dbc_init(ocelot_smart_roadster_pt)